### DISCUSSION TITLE
webcam expose set_window

### DISCUSSION THREAD

**Discussion by brianmichalk1** [arendst/Tasmota#22059] [Ideas]
In this code, 
https://github.com/espressif/esp32-camera/blob/7aa37d4f22503fdac9ccd449e4678c4894c40055/sensors/ov2640.c#L137
Is the function to set a window, which I think will set an ROI that is returned instead of the full resolution image.

First, can I submit a feature request for this to be exposed as on of the Wc commands?
Otherwise, this seems fairly easy to do myself.  What files and structs should I change?

**Comment by Jason2866** [comment]
@Staars Honestly idk if you showed me the possibility of using parts of a Webcam image only or if you have provided this as PR already.
@brianmichalk1 Before you start coding wait for the answer from Staars

**Reply by Staars** [reply]
Well, yes and no.
My last addition to Berry would allow to capture an image from the cam, get a ROI from it and process this further. Although you could in theory misuse this to present a stream of images in a browser like a normal webcam, the intention was to use it in the context of image recognition with Tensorflow.

Besides the Webcam component of Espressif would allow to define such ROI at a lower level „in the camera module“. This would probably lead to a higher performance, although IDK how much more FPS we would see.

As always it depends on the actual use case.

**Reply by brianmichalk1** [reply]
Mostly, it's about power reduction.  I've been exploring better longer focal length lenses to narrow the field of view, and use a lower resolution, Or I could select a higher pixel count and move the window around, which I think would be a better solution.
My target is a solar/battery implementation.  I've sized both for 5X bigger than I theoretically need, but even then, I think transmitting 5MP images is going to consume a lot of power.  I'm looking for a standby frame rate of around 1 fps.  If there's motion, bump that fps up higher and save to SDcard.
Using your new Berry PR may do exactly what I want.

**Reply by Staars** [reply]
I have never read any specific information/comparision regarding power consumption with different webcam modes.
But I highly doubt, that any „clever additional processing“ will reduce power consumption.
Do not forget, that we already get JPEG from the camera module so you do not have do deal with any image processing inside Tasmota at all. Just take the picture and transfer, then process it on the more powerful end device.

**Reply by brianmichalk1** [reply]
I have several applications, and one of them is to monitor water level in a cistern.  It's the most simple, and gets me into a lot of the basics I want to apply to my wireless solar solution to see if its feasible.
I can do this through my office window where I have power and such.  I'm more interested in exploring the capabilities of what I can push out to the edge, very much like your Tensor Flow conversion of digital characters.
The cistern has a big red marker on a string attached to a float.  My thought was to extract a narrow ROI, say 1000h x 100w pixels.  Average each row's red component, giving me a 1D array, apply a smoothing window, then report the peak location.  I also have this data from a pressure sensor, so I will have a nice reference to compare against.
I'm looking through your code.  I haven't yet found where you use the ROI.

**Reply by Staars** [reply]
The rough idea to work with ROIs in Berry is primarily based on static positions of ROIs in the image with a fixed cam setup.
To create the affine matrices you just take a demo picture and drag-n-drop it into the ROI editor:
https://tasmota.github.io/docs/ROI-editor/

There you create basically Berry maps, that must be converted into a byte buffer as shown here:

def roi_dsc(m)
        var d = bytes(-24)
        d.setfloat(0,m["scaleX"])
        d.setfloat(4,m["shearX"])
        d.setfloat(8,m["shearY"])
        d.setfloat(12,m["scaleY"])
        d.seti(16,m["transX"],2)
        d.seti(18,m["transY"],2)
        d.seti(20,m["width"],2)
        d.seti(22,m["height"],2)
        return d
    end
    
This buffer can be use as an arg for `img.get_buffer(roi_dsc)`.

If you need a „moving“ ROI, you have to do it in Berry, where transX and transY in the map from above are basically the pixel positions of the upper left pixel of the ROI.

**Reply by brianmichalk1** [reply]
Yes, I am reading the code in be_img_get_buffer now to see how it works.  Thanks for the help.

**Reply by brianmichalk1** [reply]
I know next to nothing about Berry.  I'm working on a simple example to get images, and the img() call is failing.  I decided to load your example, and it's also failing at the same location.  No, I don't have the Tensor Flow artifacts there.

https://tasmota.github.io/docs/TFL/#berry-implementation
line 46:        self.cam_img = img()
BRY: failed to load '/t2.be' (syntax_error - /t2.be:46: 'img' undeclared (first use in this function))

**Reply by sfromis** [reply]
To have the `img` class available, I think that you need a build with `USE_BERRY_IMAGE` defined, in the same way as already needing `USE_BERRY_TF_LITE`
